File: diag/geodsp/mk-lba-img.c
/* ----------------------------------------------------------------------- *
 *
 *   Copyright 2010 Gene Cumm
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, Inc., 53 Temple Place Ste 330,
 *   Boston MA 02111-1307, USA; either version 2 of the License, or
 *   (at your option) any later version; incorporated herein by reference.
 *
 * ----------------------------------------------------------------------- */

/*
 * mk-lba-img.c
 *
 * Makes an image that contains the LBA in every *word of every sector
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>

#define NUM_SECT (256*63+1)
#define BPS (512)
#define SECT_INT (BPS / sizeof(unsigned int))

typedef unsigned char uint8_t;
typedef unsigned int uint32_t;

const char DEF_FN[] = "-";

int main(int argc, char *argv[])
{
	int i, rv = 0, one = 0;
	unsigned int lba, b[SECT_INT];
	int len;
	FILE *f;
	uint8_t tt = 0;
	const char *fn;

	if (argc >= 2) {
		if (argc >= 3) {
			if (strcasecmp("-1", argv[1]) == 0) {
				fn = argv[2];
				one = 1;
			} else {
				fn = argv[1];
			}
		} else {
			fn = argv[1];
		}
	} else {
		fn = DEF_FN;
	}

	if (!strcmp(fn, "-"))
		f = stdout;
	else
		f = fopen(fn, "w");

	if (!f) {
		fprintf(stderr, "%s: %s: unable to open for writing: %s\n",
			argv[0], fn, strerror(errno));
		return 1;
	}

	lba = 0;
	while ((len = fread(b, 1, BPS, stdin))) {
		if (len < BPS)
			memset((char *)b + len, 0, BPS - len);
		fwrite(b, 1, BPS, f);
		lba++;
	}

	memset(b, 0, sizeof b);

	while (lba < NUM_SECT) {
		if (one) {
			b[0] = lba;
		} else {
			for (i = 0; i < SECT_INT; i++)
				b[i] = lba;
		}
		fwrite(b, 1, BPS, f);
		lba++;
	}

	if (f != stdout)
		fclose(f);

	return rv;
}

File: diag/geodsp/mk-lba-img.pl
## -----------------------------------------------------------------------
##
##   Copyright 2011 Gene Cumm
##
##   This program is free software; you can redistribute it and/or modify
##   it under the terms of the GNU General Public License as published by
##   the Free Software Foundation, Inc., 53 Temple Place Ste 330,
##   Boston MA 02111-1307, USA; either version 2 of the License, or
##   (at your option) any later version; incorporated herein by reference.
##
## -----------------------------------------------------------------------

##
## mk-lba-img.pl
##
## Make an image where each sector contains the LBA of the sector with
## a head of an input file.
##

# use bytes;

use constant SECTOR_SIZE => 512;
use constant LBA_SIZE => 8;
use constant LONG_SIZE => 4;
use constant NUM_SECTORS => (256*63+1);
# use constant NUM_SECTORS => 5;
use constant DEBUG => 1;

# sub dprint
# {
#     if (DEBUG) {
# 	print($_);
#     }
# }

($ifilen, $ofilen) = @ARGV;

if ((!defined($ifilen)) || ($ifilen eq "-")) {	# 
    print(STDERR "Using stdin\n");
    $IFILE = STDIN;
} else {
    open($IFILE, '<', $ifilen) or die "open:$!";
    print(STDERR "Using $ifilen\n");
}

binmode($ifile);

if (!defined($ofilen)) {
    $OFILE = STDOUT;
} else {
    open($OFILE, '>', $ofilen) or die "open:$!";
    print(STDERR "Using $ofilen\n");
}

binmode($OFILE);

# $pk0 = pack('L', 0);
$n_long = (SECTOR_SIZE/LONG_SIZE);
$n_lba = (SECTOR_SIZE/LBA_SIZE);

$len=0;
while ( read($IFILE, $ch, 1) ) {
    print($OFILE $ch);
    $len++;
}
$tail = (SECTOR_SIZE - ($len % SECTOR_SIZE)) % SECTOR_SIZE;
$ch = pack("C", 0);
print("Len: $len\ttail: $tail\n");
for ($i=0; $i<$tail; $i++) {
    print($OFILE $ch);
}

$st = ($len + $tail) / SECTOR_SIZE;

for ($i=$st; $i<(NUM_SECTORS); $i++) {
    @ia = ();
    for ($j=0; $j< $n_lba; $j++) {
	push(@ia, $i, 0);
    }
    @ipk = pack("L[$n_long]", @ia);
	# There is a 64-bit INT conversion but it normally isn't usable
	# on a 32-bit platform
    print($OFILE @ipk);	# Gently simulate a 64-bit LBA
}

if (defined($ifilen) && (!($ifilen eq "-"))) {
    close($IFILE);
}

if (defined($ofilen)) {
    close($OFILE);
}

exit 0;
